torch/distributed/distributed_c10d.py

```diff
@@ -474,44 +474,46 @@ def _store_based_barrier(rank, store, timeout):
     ``init_process_group`` or ``new_group``. Intended to be used only with
     those two methods and is not a generic alternative to ``barrier()``.
     """
-    store_key = "{}:{}".format(STORE_BASED_BARRIER_PREFIX, _world.group_count)
-    store.add(store_key, 1)
-    logger.info("Added key: {} to store for rank: {}".format(store_key, rank))
+    global_key = "{}:{}".format(STORE_BASED_BARRIER_PREFIX, _world.group_count)
+    def create_key(rank):
+        return "{}:{}".format(global_key, rank)
+    local_key = create_key(rank)
+    store.add(global_key, 1)
+    store.set(local_key, "1")
+    logger.info("Added key: {} to store for rank: {}".format(global_key, rank))
 
     # Now wait for all workers to check in with the store.
     world_size = get_world_size()
-    # Use 'add' instead of 'get' since for some store implementations 'add'
-    # doesn't work well with 'get'. Ideally the store implementations should
-    # be fixed, but for backward compatibility reasons it is risky to change
-    # the store implementations. Once, we completely migrate away from these
-    # legacy stores, we can use 'get' here instead.
-    worker_count = store.add(store_key, 0)
-    start = time.time()
-    log_time = time.time()
-    while worker_count != world_size:
-        time.sleep(0.01)
-        worker_count = store.add(store_key, 0)
+    all_keys = [create_key(rank) for rank in range(world_size)]
+    worker_count = store.add(global_key, 0)
 
-        # Print status periodically to keep track.
-        if timedelta(seconds=(time.time() - log_time)) > timedelta(seconds=10):
+    start = time.time()
+    while True:
+        try:
+            worker_count = store.add(global_key, 0)
+            # This will throw an exception after 10 seconds in which we print out
+            # the status of the group or time out officially, throwing runtime error
+            store.wait(all_keys, timedelta(seconds=10))
```

Comment:
This should be a global prop in c10d so we can patch it for testing purposes.

And it should be `min(10, timeout)` otherwise asking for a 1 seconds timeout will force a minimum of 10 seconds.

Reply:
to my knowledge, we don't currently have a way to validate that a log was outputted. So I am still not sure how this would be tested. We can validate the timeout works correctly by writing tests for `wait()`(i've only just done it manually right now) instead of `_store_based_barrier()`